File: engines/python/archscope_engine/common/debug_log.py
from __future__ import annotations

import json
import platform
import traceback
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional

from archscope_engine import __version__
from archscope_engine.common.redaction import (
    REDACTION_VERSION,
    merge_redaction_summaries,
    redact_text,
)

MAX_SAMPLES_PER_ERROR_TYPE = 5
MAX_CONTEXT_CHARS = 500
MAX_DEBUG_LOG_BYTES = 1_000_000

RawContext = dict[str, Optional[str]]


@dataclass
class DebugLogCollector:
    analyzer_type: str
    source_file: Path
    parser: str
    parser_options: dict[str, Any] = field(default_factory=dict)
    debug_log_dir: Path | None = None
    force_write: bool = False
    encoding_detected: str | None = None
    redaction_enabled: bool = True
    _errors: dict[str, dict[str, Any]] = field(default_factory=dict)
    _exceptions: list[dict[str, Any]] = field(default_factory=list)
    _redaction_summary: dict[str, int] = field(default_factory=dict)

    def set_redaction(
        self,
        *,
        enabled: bool = True,
        redaction_version: str = REDACTION_VERSION,
    ) -> None:
        self.redaction_enabled = enabled
        # Version is currently fixed but retained to keep the public collector API explicit.
        _ = redaction_version

    def add_parse_error(
        self,
        *,
        line_number: int | None,
        reason: str,
        message: str,
        raw_context: RawContext | None = None,
        partial_match: dict[str, Any] | None = None,
        failed_pattern: str | None = None,
        field_shapes: dict[str, Any] | None = None,
        description: str | None = None,
    ) -> None:
        entry = self._errors.setdefault(
            reason,
            {
                "count": 0,
                "description": description or message,
                "failed_pattern": failed_pattern,
                "samples": [],
            },
        )
        entry["count"] += 1
        if failed_pattern and not entry.get("failed_pattern"):
            entry["failed_pattern"] = failed_pattern
        if len(entry["samples"]) >= MAX_SAMPLES_PER_ERROR_TYPE:
            return

        redacted_context = self._redact_context(raw_context or {})
        sample = {
            "line_number": line_number,
            "raw_context": redacted_context,
            "field_shapes": field_shapes or infer_field_shapes(
                (raw_context or {}).get("target") or ""
            ),
            "partial_match": partial_match,
            "message": message,
        }
        entry["samples"].append(sample)

    def add_exception(
        self,
        *,
        phase: str,
        exception: BaseException,
        line_number: int | None = None,
        raw_context: RawContext | None = None,
    ) -> None:
        self._exceptions.append(
            {
                "phase": phase,
                "line_number": line_number,
                "exception_type": type(exception).__name__,
                "message": redact_text(str(exception)).text,
                "traceback": traceback.format_exc(),
                "raw_context": self._redact_context(raw_context or {}),
            }
        )

    def should_write(self) -> bool:
        return self.force_write or bool(self._errors) or bool(self._exceptions)

    def write(
        self,
        *,
        diagnostics: dict[str, Any] | None = None,
        output_dir: Path | None = None,
    ) -> Path | None:
        if not self.should_write():
            return None

        target_dir = output_dir or self.debug_log_dir or default_debug_log_dir()
        target_dir.mkdir(parents=True, exist_ok=True)
        path = target_dir / self.filename()
        payload = self.to_dict(diagnostics=diagnostics, debug_log_dir=target_dir)
        _write_size_limited_json(path, payload)
        return path

    def filename(self) -> str:
        timestamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
        return f"archscope-debug-{self.analyzer_type}-{timestamp}.json"

    def to_dict(
        self,
        *,
        diagnostics: dict[str, Any] | None = None,
        debug_log_dir: Path | None = None,
    ) -> dict[str, Any]:
        total_lines = _int_from_diagnostics(diagnostics, "total_lines")
        parsed = _int_from_diagnostics(diagnostics, "parsed_records")
        skipped = _int_from_diagnostics(diagnostics, "skipped_lines")
        error_types = {
            reason: int(entry["count"])
            for reason, entry in sorted(self._errors.items())
        }
        return {
            "environment": {
                "archscope_version": __version__,
                "python_version": platform.python_version(),
                "os": platform.platform(),
                "timestamp": datetime.now(timezone.utc).isoformat(),
            },
            "context": {
                "analyzer_type": self.analyzer_type,
                "source_file": redact_text(str(self.source_file)).text,
                "source_file_name": self.source_file.name,
                "file_size_bytes": _file_size(self.source_file),
                "encoding_detected": self.encoding_detected,
                "parser": self.parser,
                "parser_options": dict(self.parser_options),
                "debug_log_dir": str(
                    debug_log_dir or self.debug_log_dir or default_debug_log_dir()
                ),
            },
            "redaction": {
                "enabled": self.redaction_enabled,
                "redaction_version": REDACTION_VERSION,
                "raw_context_redacted": self.redaction_enabled,
                "summary": dict(sorted(self._redaction_summary.items())),
            },
            "summary": {
                "total_lines": total_lines,
                "parsed_ok": parsed,
                "skipped": skipped or sum(error_types.values()),
                "skip_rate_percent": _skip_rate(total_lines, skipped),
                "error_types": error_types,
                "exceptions": len(self._exceptions),
                "verdict": _verdict(total_lines, skipped, len(self._exceptions)),
            },
            "errors_by_type": {
                reason: _clean_error_entry(entry)
                for reason, entry in sorted(self._errors.items())
            },
            "exceptions": list(self._exceptions),
            "hints": _build_hints(total_lines, skipped, error_types, self._exceptions),
        }

    def _redact_context(self, raw_context: RawContext) -> RawContext:
        redacted: RawContext = {}
        summaries: list[dict[str, int]] = []
        for key in ("before", "target", "after"):
            value = raw_context.get(key)
            if value is None:
                redacted[key] = None
                continue
            clipped = value[:MAX_CONTEXT_CHARS]
            if self.redaction_enabled:
                result = redact_text(clipped)
                redacted[key] = result.text
                summaries.append(result.summary)
            else:
                redacted[key] = clipped
        self._redaction_summary = merge_redaction_summaries(
            self._redaction_summary,
            *summaries,
        )
        return redacted


def default_debug_log_dir() -> Path:
    return Path.cwd() / "archscope-debug"


def infer_field_shapes(text: str) -> dict[str, Any]:
    quote_count = text.count('"')
    shapes: dict[str, Any] = {
        "target_token_count": len(text.split()),
        "quote_count": quote_count,
        "bracket_count": text.count("[") + text.count("]"),
    }
    if " " in text and "=" in text.rsplit(" ", 1)[-1]:
        shapes["suffix_shape"] = "key=value"
    request = _extract_request_shape(text)
    if request:
        shapes.update(request)
    timestamp_shape = _timestamp_shape(text)
    if timestamp_shape:
        shapes["timestamp_shape"] = timestamp_shape
    return shapes


def _extract_request_shape(text: str) -> dict[str, Any] | None:
    import re

    match = re.search(r'"(?P<method>[A-Z]+)\s+(?P<path>\S+)\s+(?P<protocol>[^"]+)"', text)
    if not match:
        return None
    path = match.group("path")
    query_keys: list[str] = []
    if "?" in path:
        query = path.split("?", 1)[1]
        query_keys = [part.split("=", 1)[0] for part in query.split("&") if part]
    return {
        "request_shape": (
            "METHOD PATH_WITH_QUERY PROTOCOL" if query_keys else "METHOD PATH PROTOCOL"
        ),
        "path_shape": _path_shape(path.split("?", 1)[0]),
        "query_keys": query_keys,
    }


def _path_shape(path: str) -> str:
    import re

    return re.sub(r"/\d+", "/<NUMBER>", path)


def _timestamp_shape(text: str) -> str | None:
    import re

    if re.search(r"\[\d{2}/[A-Za-z]{3}/\d{4}:\d{2}:\d{2}:\d{2} [+-]\d{4}\]", text):
        return "dd/Mon/yyyy:HH:mm:ss Z"
    if re.search(r"\[\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\]", text):
        return "yyyy-MM-dd HH:mm:ss"
    if re.search(r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}", text):
        return "ISO-8601"
    return None


def _write_size_limited_json(path: Path, payload: dict[str, Any]) -> None:
    while True:
        encoded = json.dumps(payload, ensure_ascii=False, indent=2).encode("utf-8")
        if len(encoded) <= MAX_DEBUG_LOG_BYTES:
            path.write_bytes(encoded)
            return
        if _remove_oldest_sample(payload):
            continue
        _shrink_payload(payload)
        encoded = json.dumps(payload, ensure_ascii=False, indent=2).encode("utf-8")
        if len(encoded) <= MAX_DEBUG_LOG_BYTES:
            path.write_bytes(encoded)
            return
        path.write_text(
            json.dumps(_minimal_payload(payload), ensure_ascii=False, indent=2),
            encoding="utf-8",
        )
        return


def _remove_oldest_sample(payload: dict[str, Any]) -> bool:
    errors = payload.get("errors_by_type")
    if not isinstance(errors, dict):
        return False
    for entry in errors.values():
        if isinstance(entry, dict):
            samples = entry.get("samples")
            if isinstance(samples, list) and samples:
                samples.pop(0)
                return True
    return False


def _shrink_payload(payload: dict[str, Any]) -> None:
    payload["truncated"] = True
    for exception in payload.get("exceptions", []):
        if isinstance(exception, dict) and isinstance(exception.get("traceback"), str):
            exception["traceback"] = exception["traceback"][:2000]


def _minimal_payload(payload: dict[str, Any]) -> dict[str, Any]:
    return {
        "environment": payload.get("environment", {}),
        "context": payload.get("context", {}),
        "redaction": payload.get("redaction", {}),
        "summary": payload.get("summary", {}),
        "truncated": True,
        "hints": ["Debug log exceeded size cap; samples were removed."],
    }


def _int_from_diagnostics(diagnostics: dict[str, Any] | None, key: str) -> int:
    if not diagnostics:
        return 0
    value = diagnostics.get(key)
    return value if isinstance(value, int) else 0


def _skip_rate(total_lines: int, skipped: int) -> float:
    if total_lines <= 0:
        return 0.0
    return round(skipped / total_lines * 100, 2)


def _verdict(total_lines: int, skipped: int, exceptions: int) -> str:
    if exceptions > 0:
        return "FATAL_ERROR"
    if skipped <= 0:
        return "CLEAN"
    if total_lines > 0 and skipped / total_lines >= 0.5:
        return "MAJORITY_FAILED"
    return "PARTIAL_SUCCESS"


def _build_hints(
    total_lines: int,
    skipped: int,
    error_types: dict[str, int],
    exceptions: list[dict[str, Any]],
) -> list[str]:
    hints: list[str] = []
    total_errors = sum(error_types.values())
    if total_errors:
        no_format = error_types.get("NO_FORMAT_MATCH", 0)
        if no_format / total_errors >= 0.8:
            hints.append(
                "NO_FORMAT_MATCH dominates. The input may not match the selected parser format."
            )
        if error_types.get("INVALID_TIMESTAMP", 0) > 0:
            hints.append(
                "INVALID_TIMESTAMP is present. Check timestamp shape and parser time format."
            )
    if total_lines > 0 and skipped / total_lines >= 0.5:
        hints.append(
            "More than half of parsed lines failed. The file may use an unsupported format."
        )
    if exceptions:
        hints.append("Parser or analyzer exception captured. Inspect traceback and phase.")
    return hints


def _clean_error_entry(entry: dict[str, Any]) -> dict[str, Any]:
    cleaned = {
        "count": entry["count"],
        "description": entry.get("description"),
        "samples": entry.get("samples", []),
    }
    if entry.get("failed_pattern"):
        cleaned["failed_pattern"] = entry["failed_pattern"]
    return cleaned


def _file_size(path: Path) -> int | None:
    try:
        return path.stat().st_size
    except OSError:
        return None

File: apps/engine-native/internal/profiler/debuglog.go
// ─────────────────────────────────────────────────────────────────────
// [한글] debuglog — parser 실패를 portable 한 단일 JSON 로그로 수집.
//
// 책임/목적
//   parser 가 어디서 어떻게 실패했는지를 사용자가 원본 데이터를 공유하지
//   않고도 보낼 수 있도록 redacted 단일 JSON 으로 모은다. Python 측
//   archscope_engine.common.debug_log.DebugLogCollector 와 동일한 JSON
//   shape, 같은 sample cap (per-error-type 5건), 같은 1MiB output cap.
//
// 수집되는 정보
//   - environment: archscope_version / go_version / os / timestamp
//   - context    : analyzer_type / source_file (redacted) / file_size /
//                  encoding_detected / parser / parser_options
//   - redaction  : enabled / version / summary (카테고리별 redact 횟수)
//   - summary    : total_lines / parsed_ok / skipped / skip_rate /
//                  error_types / exceptions / verdict
//   - errors_by_type : reason 별 count + description + samples (cap 5)
//   - exceptions : fatal exception 들의 phase/message/traceback
//   - hints      : 자동 진단 힌트 (NO_FORMAT_MATCH 80%↑, 50%↑ skip 등)
//
// Verdict 판정 정책 (Python 과 동일)
//   - exception 있음     → FATAL_ERROR
//   - skipped >= 50%     → MAJORITY_FAILED
//   - skipped > 0        → PARTIAL_SUCCESS
//   - 그 외              → CLEAN
//
// 추가 메타
//   - InferFieldShapes : 한 라인의 토큰 수, quote 수, bracket 수,
//     suffix=key=value 여부, request shape (method/path/protocol),
//     timestamp shape (Nginx/Bracket/ISO-8601) 추론. 진단의 "왜 실패
//     했나" 단서로 활용. Python 측과 같은 키 이름 + 값 사용.
//
// 트리키한 부분
//   • per-error-type cap 은 truncated 모드에서도 5개 유지. 원본 raw 가
//     커서 전체 payload 가 1MiB 넘으면 oldest 부터 drop.
//   • errorsByType 출력 순서는 reason ASC sort (deterministic).
//   • RawContext 는 redact + 500자 절단 후 저장. token/cookie/url 등
//     민감 정보가 byte-level redact 적용된 값만 디스크로 나간다.
//   • PortableFilename 은 "archscope-debug-<parser>-<UTC stamp>.json".
//     filename component 는 [a-zA-Z0-9_-] 외 모두 "_" 로 치환.
// ─────────────────────────────────────────────────────────────────────

package profiler

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strings"
	"time"
)

// [한글] maxContextChars — RawContext 한 필드(before/target/after)의 최대 길이.
const maxContextChars = 500

// DebugLog captures a portable, redacted parser failure log that a field user
// can ship as a single artifact without sharing the original input.
//
// Mirrors archscope_engine.common.debug_log.DebugLogCollector — same JSON
// shape, same per-error-type sample cap, same 1 MiB output cap, same
// redact-by-default policy. Verdict / encoding metadata is decided at write
// time so callers don't have to know the policy.
type DebugLog struct {
	analyzerType     string
	parser           string
	sourceFile       string
	encodingDetected string
	parserOptions    map[string]any
	totalLines       int
	parsedRecords    int
	skippedLines     int
	parseErrors      []DebugLogError
	exceptions       []DebugLogException
	hints            []string
	redactionSummary map[string]int
	startedAt        time.Time
	maxSamplesByType int
	maxBytes         int
}

// DebugLogError is a single parser failure record. `RawContext` keeps a small
// before/target/after window of the offending input region (already redacted)
// so the developer can reproduce the failure without the full file.
type DebugLogError struct {
	LineNumber    int               `json:"line_number"`
	Reason        string            `json:"reason"`
	Message       string            `json:"message"`
	FailedPattern string            `json:"failed_pattern,omitempty"`
	RawContext    map[string]string `json:"raw_context,omitempty"`
	FieldShapes   map[string]any    `json:"field_shapes,omitempty"`
}

// DebugLogException records a fatal parser exception (traceback / phase) so
// a crashing run still ships actionable evidence.
type DebugLogException struct {
	Phase     string `json:"phase"`
	Message   string `json:"message"`
	Traceback string `json:"traceback,omitempty"`
}

// NewDebugLog constructs a collector. `parser` should match the parser name
// recorded under `metadata.parser` in the AnalysisResult so the log can be
// correlated with the analyzer output.
func NewDebugLog(analyzerType, parser, sourceFile string) *DebugLog {
	return &DebugLog{
		analyzerType:     analyzerType,
		parser:           parser,
		sourceFile:       sourceFile,
		parserOptions:    map[string]any{},
		parseErrors:      []DebugLogError{},
		exceptions:       []DebugLogException{},
		hints:            []string{},
		redactionSummary: map[string]int{},
		startedAt:        time.Now().UTC(),
		maxSamplesByType: 5,
		maxBytes:         1024 * 1024,
	}
}

// SetEncodingDetected records the text encoding the parser settled on.
func (l *DebugLog) SetEncodingDetected(encoding string) {
	if l == nil {
		return
	}
	l.encodingDetected = encoding
}

// SetTotals records the final parsing totals for the summary section.
func (l *DebugLog) SetTotals(totalLines, parsedRecords, skippedLines int) {
	if l == nil {
		return
	}
	l.totalLines = totalLines
	l.parsedRecords = parsedRecords
	l.skippedLines = skippedLines
}

// SetParserOptions records the parser option snapshot.
func (l *DebugLog) SetParserOptions(opts map[string]any) {
	if l == nil || opts == nil {
		return
	}
	l.parserOptions = opts
}

// AddHint adds a free-form developer hint shown next to the verdict.
func (l *DebugLog) AddHint(hint string) {
	if l == nil || strings.TrimSpace(hint) == "" {
		return
	}
	l.hints = append(l.hints, hint)
}

// AddParseError records a row-level parser failure. `rawContext` may include
// `before`, `target`, `after` strings; each is redacted in place. The total
// per-`reason` sample count is capped at `maxSamplesByType` so a failing
// 100k-line input does not bloat the log.
//
// [한글] AddParseError — row-level 파싱 실패 1건 기록.
// 같은 reason 으로 이미 5건 쌓였으면 추가 기록 없이 즉시 반환 (cap).
// rawContext 의 before/target/after 는 각각 500자 절단 + RedactText 적용.
// fieldShapes 가 비어있으면 target 으로부터 InferFieldShapes 자동 추론.
func (l *DebugLog) AddParseError(lineNumber int, reason, message, failedPattern string, rawContext map[string]string, fieldShapes map[string]any) {
	if l == nil {
		return
	}
	count := 0
	for _, existing := range l.parseErrors {
		if existing.Reason == reason {
			count++
		}
	}
	if count >= l.maxSamplesByType {
		return
	}
	redactedContext := map[string]string{}
	for _, key := range []string{"before", "target", "after"} {
		value, ok := rawContext[key]
		if !ok {
			continue
		}
		clipped := value
		if len(clipped) > maxContextChars {
			clipped = clipped[:maxContextChars]
		}
		result := RedactText(clipped)
		redactedContext[key] = result.Text
		l.redactionSummary = MergeRedactionSummaries(l.redactionSummary, result.Summary)
	}
	if fieldShapes == nil {
		target := rawContext["target"]
		if target != "" {
			fieldShapes = InferFieldShapes(target)
		}
	}
	l.parseErrors = append(l.parseErrors, DebugLogError{
		LineNumber:    lineNumber,
		Reason:        reason,
		Message:       message,
		FailedPattern: failedPattern,
		RawContext:    redactedContext,
		FieldShapes:   fieldShapes,
	})
}

// AddException records a fatal exception during analysis.
func (l *DebugLog) AddException(phase, message, traceback string) {
	if l == nil {
		return
	}
	l.exceptions = append(l.exceptions, DebugLogException{
		Phase:     phase,
		Message:   message,
		Traceback: traceback,
	})
}

// HasContent reports whether the collector picked up anything worth writing.
func (l *DebugLog) HasContent() bool {
	if l == nil {
		return false
	}
	return len(l.parseErrors) > 0 || len(l.exceptions) > 0
}

// Verdict ranks the log using the same policy as Python:
// FATAL_ERROR if exceptions, MAJORITY_FAILED if ≥50% skipped,
// PARTIAL_SUCCESS if any skipped, CLEAN otherwise.
func (l *DebugLog) Verdict() string {
	if l == nil {
		return "CLEAN"
	}
	if len(l.exceptions) > 0 {
		return "FATAL_ERROR"
	}
	skipped := l.skippedLines
	if skipped <= 0 {
		// Fall back to counting parse errors if SetTotals wasn't called.
		skipped = len(l.parseErrors)
	}
	if skipped <= 0 {
		return "CLEAN"
	}
	if l.totalLines > 0 && float64(skipped)/float64(l.totalLines) >= 0.5 {
		return "MAJORITY_FAILED"
	}
	return "PARTIAL_SUCCESS"
}

// PortableFilename returns a deterministic file name suitable for shipping.
func (l *DebugLog) PortableFilename() string {
	stem := "archscope-debug"
	if l != nil && l.parser != "" {
		stem = stem + "-" + sanitizeFilenameComponent(l.parser)
	}
	stamp := time.Now().UTC().Format("20060102T150405Z")
	return fmt.Sprintf("%s-%s.json", stem, stamp)
}

func sanitizeFilenameComponent(value string) string {
	out := []rune{}
	for _, r := range value {
		switch {
		case r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z', r >= '0' && r <= '9', r == '-', r == '_':
			out = append(out, r)
		default:
			out = append(out, '_')
		}
	}
	return string(out)
}

// WriteJSON serializes the collected log into the configured directory. If
// `dir` is empty, defaults to `<execution-cwd>/archscope-debug/`. Returns the
// final output path. The payload is JSON-encoded and capped at `maxBytes`;
// when the cap kicks in the most recent samples are kept and `truncated:
// true` is emitted in the metadata.
//
// [한글] WriteJSON — payload(비truncate) 직렬화 → maxBytes(1MiB) 초과면
// truncated payload 로 재직렬화 → 0644 로 디스크 기록. 디렉토리 자동 생성.
// 반환값은 실제 기록된 파일 경로 (filename 은 PortableFilename).
func (l *DebugLog) WriteJSON(dir string) (string, error) {
	if l == nil {
		return "", fmt.Errorf("debug log not initialized")
	}
	if dir == "" {
		cwd, err := os.Getwd()
		if err != nil {
			return "", err
		}
		dir = filepath.Join(cwd, "archscope-debug")
	}
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return "", err
	}
	path := filepath.Join(dir, l.PortableFilename())

	payload := l.payload(false)
	body, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		return "", err
	}
	if len(body) > l.maxBytes {
		// Fall back to a truncated payload — drop oldest samples first.
		payload = l.payload(true)
		body, err = json.MarshalIndent(payload, "", "  ")
		if err != nil {
			return "", err
		}
	}
	if err := os.WriteFile(path, body, 0o644); err != nil {
		return "", err
	}
	return path, nil
}

func (l *DebugLog) payload(truncated bool) map[string]any {
	// Group parse errors by reason, matching Python's errors_by_type shape.
	type errorEntry struct {
		Count         int              `json:"count"`
		Description   string           `json:"description"`
		FailedPattern string           `json:"failed_pattern,omitempty"`
		Samples       []DebugLogError  `json:"samples"`
	}
	errorsByType := map[string]*errorEntry{}
	for _, err := range l.parseErrors {
		entry, ok := errorsByType[err.Reason]
		if !ok {
			entry = &errorEntry{Description: err.Message, FailedPattern: err.FailedPattern}
			errorsByType[err.Reason] = entry
		}
		entry.Count++
		if entry.FailedPattern == "" && err.FailedPattern != "" {
			entry.FailedPattern = err.FailedPattern
		}
		if truncated && len(entry.Samples) >= l.maxSamplesByType {
			continue
		}
		if len(entry.Samples) < l.maxSamplesByType {
			entry.Samples = append(entry.Samples, err)
		}
	}
	// Sorted by reason for deterministic output.
	sortedReasons := make([]string, 0, len(errorsByType))
	for reason := range errorsByType {
		sortedReasons = append(sortedReasons, reason)
	}
	sort.Strings(sortedReasons)
	orderedErrorsByType := map[string]any{}
	errorTypeCounts := map[string]int{}
	for _, reason := range sortedReasons {
		entry := errorsByType[reason]
		errorTypeCounts[reason] = entry.Count
		cleaned := map[string]any{
			"count":       entry.Count,
			"description": entry.Description,
			"samples":     entry.Samples,
		}
		if entry.FailedPattern != "" {
			cleaned["failed_pattern"] = entry.FailedPattern
		}
		orderedErrorsByType[reason] = cleaned
	}

	skipped := l.skippedLines
	if skipped <= 0 {
		for _, entry := range errorsByType {
			skipped += entry.Count
		}
	}
	skipRate := 0.0
	if l.totalLines > 0 {
		skipRate = float64(skipped) / float64(l.totalLines) * 100
		skipRate = float64(int(skipRate*100)) / 100 // round to 2 decimals
	}

	redactedSourceFile := RedactText(l.sourceFile).Text
	sourceFileName := filepath.Base(l.sourceFile)

	var fileSize *int64
	if info, err := os.Stat(l.sourceFile); err == nil {
		s := info.Size()
		fileSize = &s
	}

	return map[string]any{
		"environment": map[string]string{
			"archscope_version": "go-native",
			"go_version":        runtime.Version(),
			"os":                runtime.GOOS + "/" + runtime.GOARCH,
			"timestamp":         time.Now().UTC().Format(time.RFC3339),
		},
		"context": map[string]any{
			"analyzer_type":    l.analyzerType,
			"source_file":      redactedSourceFile,
			"source_file_name": sourceFileName,
			"file_size_bytes":  fileSize,
			"encoding_detected": l.encodingDetected,
			"parser":            l.parser,
			"parser_options":    l.parserOptions,
		},
		"redaction": map[string]any{
			"enabled":              true,
			"redaction_version":    RedactionVersion,
			"raw_context_redacted": true,
			"summary":              l.redactionSummary,
		},
		"summary": map[string]any{
			"total_lines":      l.totalLines,
			"parsed_ok":        l.parsedRecords,
			"skipped":          skipped,
			"skip_rate_percent": skipRate,
			"error_types":      errorTypeCounts,
			"exceptions":       len(l.exceptions),
			"verdict":          l.Verdict(),
		},
		"errors_by_type": orderedErrorsByType,
		"exceptions":     l.exceptions,
		"hints":          l.buildHints(skipped, errorTypeCounts),
		"truncated":      truncated,
	}
}

// buildHints generates developer hints matching Python's _build_hints.
func (l *DebugLog) buildHints(skipped int, errorTypes map[string]int) []string {
	hints := append([]string{}, l.hints...)
	totalErrors := 0
	for _, count := range errorTypes {
		totalErrors += count
	}
	if totalErrors > 0 {
		noFormat := errorTypes["NO_FORMAT_MATCH"]
		if float64(noFormat)/float64(totalErrors) >= 0.8 {
			hints = append(hints, "NO_FORMAT_MATCH dominates. The input may not match the selected parser format.")
		}
		if errorTypes["INVALID_TIMESTAMP"] > 0 {
			hints = append(hints, "INVALID_TIMESTAMP is present. Check timestamp shape and parser time format.")
		}
	}
	if l.totalLines > 0 && float64(skipped)/float64(l.totalLines) >= 0.5 {
		hints = append(hints, "More than half of parsed lines failed. The file may use an unsupported format.")
	}
	if len(l.exceptions) > 0 {
		hints = append(hints, "Parser or analyzer exception captured. Inspect traceback and phase.")
	}
	return hints
}

// InferFieldShapes extracts structural metadata from a raw text line,
// matching Python's infer_field_shapes for cross-engine parity.
//
// [한글] InferFieldShapes — raw 라인 한 줄에서 구조적 단서 추출.
// 토큰 수, quote/bracket 카운트, suffix=key=value 패턴, request shape
// (method/path/protocol), timestamp shape 등을 추론해 진단의 "왜 실패
// 했나" 단서로 사용. Python 측 infer_field_shapes 와 byte-level 동등.
func InferFieldShapes(text string) map[string]any {
	shapes := map[string]any{
		"target_token_count": len(strings.Fields(text)),
		"quote_count":        strings.Count(text, "\""),
		"bracket_count":      strings.Count(text, "[") + strings.Count(text, "]"),
	}
	// Suffix key=value detection.
	if idx := strings.LastIndex(text, " "); idx >= 0 {
		suffix := text[idx+1:]
		if strings.Contains(suffix, "=") {
			shapes["suffix_shape"] = "key=value"
		}
	}
	// Request shape detection.
	if reqShape := extractRequestShape(text); reqShape != nil {
		for k, v := range reqShape {
			shapes[k] = v
		}
	}
	// Timestamp shape detection.
	if ts := detectTimestampShape(text); ts != "" {
		shapes["timestamp_shape"] = ts
	}
	return shapes
}

var (
	requestShapeRE      = regexp.MustCompile(`"(?P<method>[A-Z]+)\s+(?P<path>\S+)\s+(?P<protocol>[^"]+)"`)
	timestampNginxRE    = regexp.MustCompile(`\[\d{2}/[A-Za-z]{3}/\d{4}:\d{2}:\d{2}:\d{2} [+-]\d{4}\]`)
	timestampBracketRE  = regexp.MustCompile(`\[\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\]`)
	timestampISO8601RE  = regexp.MustCompile(`\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}`)
	pathNumberRE        = regexp.MustCompile(`/\d+`)
)

func extractRequestShape(text string) map[string]any {
	match := requestShapeRE.FindStringSubmatch(text)
	if match == nil {
		return nil
	}
	path := match[2]
	queryKeys := []string{}
	if idx := strings.Index(path, "?"); idx >= 0 {
		query := path[idx+1:]
		for _, part := range strings.Split(query, "&") {
			if part == "" {
				continue
			}
			kv := strings.SplitN(part, "=", 2)
			queryKeys = append(queryKeys, kv[0])
		}
	}
	result := map[string]any{}
	if len(queryKeys) > 0 {
		result["request_shape"] = "METHOD PATH_WITH_QUERY PROTOCOL"
	} else {
		result["request_shape"] = "METHOD PATH PROTOCOL"
	}
	purePath := path
	if idx := strings.Index(purePath, "?"); idx >= 0 {
		purePath = purePath[:idx]
	}
	result["path_shape"] = pathNumberRE.ReplaceAllString(purePath, "/<NUMBER>")
	result["query_keys"] = queryKeys
	return result
}

func detectTimestampShape(text string) string {
	if timestampNginxRE.MatchString(text) {
		return "dd/Mon/yyyy:HH:mm:ss Z"
	}
	if timestampBracketRE.MatchString(text) {
		return "yyyy-MM-dd HH:mm:ss"
	}
	if timestampISO8601RE.MatchString(text) {
		return "ISO-8601"
	}
	return ""
}
